registry/remote/auth/example_test.go

```diff
@@ -0,0 +1,219 @@
+			authHeader := r.Header.Get("Authorization")
+			if authHeader != wantedAuthHeader {
+				w.Header().Set("Www-Authenticate", `Basic realm="Test Server"`)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/simple":
+			w.WriteHeader(http.StatusOK)
+		default:
+			w.WriteHeader(http.StatusNotAcceptable)
+		}
+	}))
+	defer ts.Close()
+	host = ts.URL
+	simpleURL = fmt.Sprintf("%s/simple", host)
+	basicAuthURL = fmt.Sprintf("%s/basicAuth", host)
+	accessTokenURL = fmt.Sprintf("%s/accessToken", host)
+	clientConfigURL = fmt.Sprintf("%s/clientConfig", host)
+	http.DefaultClient = ts.Client()
+
+	os.Exit(m.Run())
+}
+
+// ExampleClient_Do_minimalClient gives an example of a minimal working client.
+func ExampleClient_Do_minimalClient() {
+	var client auth.Client
+	req, err := http.NewRequest(http.MethodGet, simpleURL, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_basicAuth gives an example of using client with credentials.
+func ExampleClient_Do_basicAuth() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			switch reg {
+			case host[8:]:
+				return auth.Credential{
+					Username: username,
+					Password: password,
+				}, nil
+			default:
+				return auth.EmptyCredential, fmt.Errorf("credential not found for %v", reg)
+			}
+		},
+	}
+	req, err := http.NewRequest(http.MethodGet, basicAuthURL, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_withAccessToken gives an example of using client with an access token.
+func ExampleClient_Do_withAccessToken() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			return auth.Credential{
+				AccessToken: accessToken,
+			}, nil
+		},
+	}
+	req, err := http.NewRequest(http.MethodGet, accessTokenURL, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_clientConfiguration shows the client configurations available,
+// including using cache, setting user agent and configuring OAuth2.
+func ExampleClient_Do_clientConfiguration() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			switch reg {
+			case host[8:]:
+				return auth.Credential{
+					Username: username,
+					Password: password,
+				}, nil
+			default:
+				return auth.EmptyCredential, fmt.Errorf("credential not found for %v", reg)
+			}
+		},
+		ForceAttemptOAuth2: false,
+		Cache:              auth.NewCache(),
+	}
+	client.SetUserAgent("example user agent")
+	ctx := auth.WithScopes(context.Background(), scopes...)
```

Comment:
You might need more doc on explaining the line 202 and line 205. Also, users reading this example will be confused on what `scopes` is.

Reply:
Added relevant comments to explain.